uefi/src/proto/network/http.rs

```diff
@@ -0,0 +1,370 @@
+use uefi_raw::protocol::network::http::{
+    HttpAccessPoint, HttpConfigData, HttpHeader, HttpMessage, HttpMethod, HttpProtocol,
+    HttpRequestData, HttpResponseData, HttpStatusCode, HttpToken, HttpV4AccessPoint, HttpVersion,
+};
+
+/// HTTP Protocol
+#[derive(Debug)]
+#[unsafe_protocol(HttpProtocol::GUID)]
+pub struct Http(HttpProtocol);
+
+impl Http {
+    /// Configure HTTP Protocol.  Must be called before sending HTTP requests.
+    pub fn configure(&mut self, config_data: &HttpConfigData) -> uefi::Result<()> {
+        let status = unsafe { (self.0.configure)(&mut self.0, config_data) };
+        match status {
+            Status::SUCCESS => Ok(()),
+            _ => Err(status.into()),
+        }
+    }
+
+    /// Send HTTP request.
+    pub fn request(&mut self, token: &mut HttpToken) -> uefi::Result<()> {
+        let status = unsafe { (self.0.request)(&mut self.0, token) };
+        match status {
+            Status::SUCCESS => Ok(()),
+            _ => Err(status.into()),
+        }
+    }
+
+    /// Receive HTTP response.
+    pub fn response(&mut self, token: &mut HttpToken) -> uefi::Result<()> {
+        let status = unsafe { (self.0.response)(&mut self.0, token) };
+        match status {
+            Status::SUCCESS => Ok(()),
+            _ => Err(status.into()),
+        }
+    }
+
+    /// Poll network stack for updates.
+    pub fn poll(&mut self) -> uefi::Result<()> {
+        let status = unsafe { (self.0.poll)(&mut self.0) };
+        match status {
+            Status::SUCCESS => Ok(()),
+            _ => Err(status.into()),
+        }
+    }
+}
+
+/// HTTP Service Binding Protocol.
+#[derive(Debug)]
+#[unsafe_protocol(HttpProtocol::SERVICE_BINDING_GUID)]
+pub struct HttpBinding(ServiceBindingProtocol);
+
+impl HttpBinding {
+    /// Create HTTP Protocol Handle.
+    pub fn create_child(&mut self) -> uefi::Result<Handle> {
+        let mut c_handle = ptr::null_mut();
+        let status;
+        let handle;
+        unsafe {
+            status = (self.0.create_child)(&mut self.0, &mut c_handle);
+            handle = Handle::from_ptr(c_handle);
+        };
+        match status {
+            Status::SUCCESS => Ok(handle.unwrap()),
+            _ => Err(status.into()),
+        }
+    }
+
+    /// Destroy HTTP Protocol Handle.
+    pub fn destroy_child(&mut self, handle: Handle) -> uefi::Result<()> {
+        let status = unsafe { (self.0.destroy_child)(&mut self.0, handle.as_ptr()) };
+        match status {
+            Status::SUCCESS => Ok(()),
+            _ => Err(status.into()),
+        }
+    }
+}
+
+/// HTTP Response data
+#[derive(Debug)]
+pub struct HttpHelperResponse {
+    /// HTTP Status
+    pub status: HttpStatusCode,
+    /// HTTP Response Headers
+    pub headers: Vec<(String, String)>,
+    /// HTTP Body
+    pub body: Vec<u8>,
+}
+
+/// HTTP Helper, makes using the HTTP protocol more convinient.
+#[derive(Debug)]
+pub struct HttpHelper {
+    child_handle: Option<Handle>,
```

Comment:
it feels weird that all fields can be `None`. We might need a `Builder` for that?

Reply:
IIRC the reason for that was to create the HttpHelper object at the start of the new() function and have the drop trait cleanup properly (specifically call destroy_child) in case I have a failure half way through.

Reworking the error handling and not allowing half-initialized HttpHelper objects exist is probably a good idea.